File: web/lib/metrics.ts
/**
 * How the usage and quality dashboard is worded (M5 slice 2).
 *
 * It lives in `lib/` for the reason `lib/evidence.ts` does: `web/` has vitest and no
 * DOM, so a rule expressed inside a component is a rule no test can reach. Everything
 * here is a pure function of the server's payload.
 *
 * The distinction this module exists to protect: **"we spent nothing" and "we do not
 * know what we spent" are different sentences, and only one of them is safe to render
 * as a number.** `cost_usd` is null when any call in a group has no known price, and
 * showing that as `$0.00` would be the same silent corruption as a stale price table —
 * a hazard `CLAUDE.md` names. Today every call is free-tier and genuinely costs 0.00,
 * which is exactly when the two are easiest to confuse and the confusion is invisible.
 */

import type {
  CallBucket,
  CallTotals,
  ParseOutcome,
  QualitySummary,
  UsageReport,
  UsageScope,
} from "@/lib/api";

/** What a figure reads as when the data cannot support one. */
export const UNKNOWN = "unknown";

/** What a figure reads as when there is simply nothing yet. */
export const NOTHING_YET = "—";

/**
 * Money, or an honest refusal to name it.
 *
 * Never returns "$0.00" for an unknown cost. That is the entire job.
 */
export function formatCost(totals: CallTotals): string {
  if (totals.cost_usd === null) return UNKNOWN;
  return `$${totals.cost_usd.toFixed(4)}`;
}

/**
 * Why a cost is unknown, in words, or null when it is not.
 *
 * A bare "unknown" invites the reader to assume a bug. Naming the rows behind it —
 * the organizing idea of this milestone — makes it a measurement instead.
 */
export function describeCostGap(totals: CallTotals): string | null {
  if (totals.cost_usd !== null) return null;
  const unpriced = totals.calls - totals.calls_priced;
  return `${unpriced} of ${totals.calls} ${totals.calls === 1 ? "call has" : "calls have"} no known price`;
}

/** Whether every call behind a figure carried a price. */
export function costIsComplete(totals: CallTotals): boolean {
  return totals.cost_usd !== null;
}

export function formatTokens(count: number): string {
  return count.toLocaleString("en-US");
}

/** Total tokens in and out. Cached input is counted in `input_tokens` already. */
export function totalTokens(totals: CallTotals): number {
  return totals.input_tokens + totals.output_tokens;
}

/** Mean latency, or a dash — the mean of zero calls is not zero. */
export function formatLatency(ms: number | null): string {
  if (ms === null) return NOTHING_YET;
  if (ms < 1000) return `${Math.round(ms)} ms`;
  return `${(ms / 1000).toFixed(1)} s`;
}

/**
 * The hallucination rate as a percentage, or a dash when no claims exist.
 *
 * `0.0%` over zero claims would read as "nothing was fabricated", which is a claim
 * about data that is not there — the same mistake as a cost of $0.00 for an unknown
 * price, one column over.
 */
export function formatRate(rate: number | null): string {
  if (rate === null) return NOTHING_YET;
  return `${(rate * 100).toFixed(1)}%`;
}

/** Whether a quality summary came through with nothing dropped. */
export function isClean(quality: QualitySummary): boolean {
  return quality.claims_dropped === 0;
}

/**
 * How many claims were refused, or null when none were.
 *
 * Here rather than inline in the component for the reason everything else is: the
 * first version read `${n} claims dropped` and rendered "1 claims dropped" on the
 * screen, which no test could see because the sentence lived inside JSX.
 */
export function droppedNote(quality: QualitySummary): string | null {
  const dropped = quality.claims_dropped;
  if (dropped === 0) return null;
  return `${formatTokens(dropped)} ${dropped === 1 ? "claim" : "claims"} dropped`;
}

const BUCKET_LABEL: Record<CallBucket, string> = {
  extraction: "Extracting profiles",
  judging: "Judging against requirements",
  unattributed: "Unattributed",
  ambiguous: "Ambiguous",
};

export function bucketLabel(bucket: CallBucket): string {
  return BUCKET_LABEL[bucket];
}

const BUCKET_NOTE: Record<CallBucket, string | null> = {
  extraction: "Billed to the resume — what it cost to read a document.",
  judging: "Billed to the screening — what it cost to judge one against a job.",
  unattributed:
    "These calls name neither a resume nor a screening, so nobody owns them. They are shown because a row that cannot be attributed would otherwise vanish from every total.",
  ambiguous:
    "These calls name both a resume and a screening, so they would be counted twice. Shown for the same reason as the row above.",
};

export function bucketNote(bucket: CallBucket): string | null {
  return BUCKET_NOTE[bucket];
}

/** The two buckets that should always be empty. Non-zero means an invariant slipped. */
export const UNEXPECTED_BUCKETS: CallBucket[] = ["unattributed", "ambiguous"];

/**
 * Whether every call could be attributed to the work that paid for it.
 *
 * `llm_call_logs` sets `resume_id` xor `screening_id`, but nothing enforces it — both
 * columns are nullable with no CHECK. So this can be false on real data, and when it
 * is, the reader needs to know before trusting any split below it.
 */
export function attributionIsClean(report: UsageReport): boolean {
  return UNEXPECTED_BUCKETS.every((bucket) => report.by_bucket[bucket].calls === 0);
}

/** The sentence to show when it is not clean, or null when it is. */
export function describeAttributionGap(report: UsageReport): string | null {
  if (attributionIsClean(report)) return null;
  const parts = UNEXPECTED_BUCKETS.filter((bucket) => report.by_bucket[bucket].calls > 0).map(
    (bucket) => `${report.by_bucket[bucket].calls} ${bucketLabel(bucket).toLowerCase()}`,
  );
  return `${parts.join(" and ")}. Every call should name exactly one of a resume or a screening.`;
}

/**
 * The buckets worth rendering: the two real ones always, the two anomalies only when
 * they are non-empty.
 *
 * Hiding an empty anomaly is not the same as hiding a blocked action — there is no
 * decision here for a reader to be confused by, and a permanent pair of zeroes trains
 * people to skip the row that matters on the day it is not zero.
 */
export function bucketsToShow(report: UsageReport): CallBucket[] {
  const always: CallBucket[] = ["extraction", "judging"];
  return [...always, ...UNEXPECTED_BUCKETS.filter((b) => report.by_bucket[b].calls > 0)];
}

const SCOPE_NOTE: Record<UsageScope, string> = {
  own: "Your own rows only — the documents you uploaded and the screenings your postings paid for.",
  all: "Every row in the system, because this account is an admin.",
};

export function scopeNote(scope: UsageScope): string {
  return SCOPE_NOTE[scope];
}

/** How many documents reached a verified profile, over how many are accounted for. */
export function parseSuccess(outcomes: ParseOutcome[]): { extracted: number; total: number } {
  const total = outcomes.reduce((sum, outcome) => sum + outcome.resumes, 0);
  const extracted = outcomes.find((outcome) => outcome.status === "extracted")?.resumes ?? 0;
  return { extracted, total };
}

/** Parse success as a percentage, or a dash when nothing has been uploaded. */
export function formatParseSuccess(outcomes: ParseOutcome[]): string {
  const { extracted, total } = parseSuccess(outcomes);
  if (total === 0) return NOTHING_YET;
  return `${((extracted / total) * 100).toFixed(0)}%`;
}

/** Whether the report has anything at all to show. */
export function isEmpty(report: UsageReport): boolean {
  return report.totals.calls === 0 && report.quality.profiles === 0;
}

File: api/app/schemas/metrics.py
"""What the usage and quality dashboard reports, and the rules it reports under.

M5 slice 2. The organizing idea, the same shape as M3's "a verdict is derived from a
located quote" and M4's "a state is a projection of an event log": **every number here
is a query over rows the system already wrote, and can name the rows it came from.**
Cite your source, applied to metrics.

So every aggregate below carries the row count behind it. A figure whose denominator is
invisible is a claim without a citation, which is the thing this project exists not to
produce — and it is what makes "42 calls" and "42 calls, of which 3 had no price"
different statements rather than the same one rounded differently.

**Nothing here spends a model call.** The dashboard reports; it never re-asks. The
schema was shaped for exactly this in M1 — `extracted_profiles` lifts
`claims_verified`, `claims_dropped` and `hallucination_rate` into real columns
*specifically* so this is a `GROUP BY` and not a JSON walk, and `models/core.py` has
said so in a docstring since then.

**This was specified as a *cost* dashboard, and was respecified with the owner on
2026-08-15 because there is no cost.** `app/llm/gemini.py` maps every model to
`FREE_TIER`, so all logged calls store `cost_usd = 0.0` and not one is `NULL` — the
single behaviour the original spec said it had to get right is unreachable on real
data. Charting it would have shipped a screen of zeroes that reads as a bug. The cost
*rule* survives here in full (see `CallTotals.cost_usd`) against the day a paid
provider lands; what the screen leads with is tokens, latency and quality, which are
real today.
"""

from __future__ import annotations

from datetime import datetime
from enum import StrEnum

from pydantic import BaseModel, Field

from app.models import ResumeStatus


class CallBucket(StrEnum):
    """Which piece of work paid for a model call.

    `LLMCallLog` sets `resume_id` **xor** `screening_id` — a call belongs to the work
    that paid for it, and a screening is not a resume. Keeping the two apart is what
    makes "what did extracting this document cost" and "what did screening this
    candidate cost" separately answerable, so the dashboard must never collapse them.

    The last two buckets exist because **that invariant is a docstring, not a
    constraint** (found 2026-08-15): both columns are nullable and nothing enforces the
    xor, so a row with neither — or with both — is legal today. Bucketing them
    explicitly is the same instinct as `dropped` and `excluded`: a row the system
    cannot attribute is *reported*, never silently absent from a total. Together the
    four buckets partition the rows exactly, which `test_metrics.py` asserts.
    """

    EXTRACTION = "extraction"
    """`resume_id` set, `screening_id` null. Building a profile from a document."""

    JUDGING = "judging"
    """`screening_id` set, `resume_id` null. Judging a document against requirements."""

    UNATTRIBUTED = "unattributed"
    """Neither set. Cannot happen through any shipped path, and would vanish from every
    owner-scoped total if it were not named — which is why it is named."""

    AMBIGUOUS = "ambiguous"
    """Both set. Also unreachable today, and it would be *double* counted rather than
    lost, so it is worth telling apart from the row that is merely missing."""


class CallTotals(BaseModel):
    """One group of model calls, summed.

    Used both per bucket and per `(provider, model, prompt_version)` group, because the
    same question — what did these calls consume — is being asked at two grains.
    """

    calls: int = Field(description="How many rows produced every figure in this group.")

    input_tokens: int
    output_tokens: int
    cached_input_tokens: int

    latency_ms_total: int
    latency_ms_mean: float | None = Field(
        default=None,
        description="Null when there are no calls, rather than a zero that reads as 'instant'.",
    )

    calls_priced: int = Field(
        description=(
            "How many of `calls` carry a known `cost_usd`. Exposed rather than implied, "
            "so a reader can see why `cost_usd` is null."
        )
    )

    cost_usd: float | None = Field(
        default=None,
        description=(
            "Null unless EVERY call in the group has a known price. SQL SUM() skips "
            "nulls, so summing a group where some prices are unknown yields a number "
            "that looks complete and is not — the same class of silent corruption as a "
            "stale price table, which CLAUDE.md names as a hazard. A partial total is "
            "worse than no total, because only one of them is obviously missing."
        ),
    )

    @property
    def cost_is_complete(self) -> bool:
        return self.calls_priced == self.calls


class CallGroup(BaseModel):
    """Calls grouped by who served them and which prompt asked.

    `prompt_version` is the axis that makes this worth showing: comparing prompt
    revisions is guesswork without it, which is why `LLMCallLog` stores it at all.
    """

    provider: str
    model: str
    prompt_version: str
    bucket: CallBucket
    totals: CallTotals


class QualitySummary(BaseModel):
    """The guardrail's own numbers, over stored profiles.

    This is the metric that costs nothing to produce — no labelled dataset, no baseline
    to beat — because verification already happens on every document.
    """

    profiles: int = Field(description="How many extracted profiles are behind these figures.")
    claims_verified: int
    claims_dropped: int

    hallucination_rate: float | None = Field(
        default=None,
        description=(
            "Recomputed from the totals — dropped / (verified + dropped) — and NOT the "
            "mean of the stored per-profile rates. Averaging rates weights a one-claim "
            "document the same as a thirty-claim one, which is a different statistic "
            "wearing this one's name. Null when there are no claims at all, rather "
            "than a 0.0 that would read as 'nothing was fabricated'."
        ),
    )

    extraction_attempts_total: int = Field(
        description=(
            "Model calls the re-ask loop spent, summed from `extracted_profiles.attempts`. "
            "This is the *stats* counter, not `Resume.attempts`, which is the job counter "
            "and means something else entirely despite the identical spelling."
        )
    )


class ParseOutcome(BaseModel):
    """How many resumes rest in each status. Parse success, without a new table."""

    status: ResumeStatus
    resumes: int


class UsageScope(StrEnum):
    """Whose rows a report covers."""

    OWN = "own"
    """The caller's own. Export is a subject-access request rather than a dump of
    everything visible, and the same instinct applies to a metric: a recruiter may
    *read* an applicant's resume without that document's extraction cost being theirs
    to total."""

    ALL = "all"
    """Every row in the system. `ADMIN` only, decided with the owner on 2026-08-15.

    Note this is a *row scope*, not a route gate, and the two are not interchangeable:
    `require_role` runs no query and returns the caller unchanged, so it can refuse a
    route but cannot widen a query. The role is therefore read here, in the WHERE
    clause, the way `_owned_resume` reads it."""


class UsageReport(BaseModel):
    """Everything the dashboard renders, in one response.

    One request rather than one per panel, for the reason `GET /jobs/{id}/ranking`
    returns verdicts *with* their citations: a screen that needs a second request per
    figure invites a screen that shows the figure without it.
    """

    scope: UsageScope
    generated_at: datetime

    totals: CallTotals = Field(
        description="Every model call in scope, whatever bucket it landed in."
    )
    by_bucket: dict[CallBucket, CallTotals] = Field(
        description=(
            "Totals per bucket. Every bucket is present, including the empty ones — an "
            "absent key and a zero are different claims, and only one of them says "
            "'we looked'."
        )
    )
    by_group: list[CallGroup] = Field(
        description="Provider, model and prompt version, ordered by call count descending."
    )

    quality: QualitySummary
    parse_outcomes: list[ParseOutcome] = Field(
        description="Ordered by count descending. Statuses with no rows are omitted."
    )
